sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -569,10 +613,353 @@ public boolean getValidate() {
+      public BoundedReader<TableRow> createReader(PipelineOptions options) throws IOException {
+        BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+        if (!Strings.isNullOrEmpty(query)) {
+          return new BigQueryReader(this, bqServices.getReaderFromQuery(
+              bqOptions, query, bqOptions.getProject(), flattenResults));
+        } else {
+          TableReference tableRef = JSON_FACTORY.fromString(jsonTable, TableReference.class);
+          return new BigQueryReader(this, bqServices.getReaderFromTable(bqOptions, tableRef));
+        }
+      }
+
+      @Override
+      public void validate() {
+        // Do nothing, validation is done in BigQuery.Read.
+      }
+
+      @Override
+      public Coder<TableRow> getDefaultOutputCoder() {
+        return TableRowJsonCoder.of();
+      }
+
+      private static void executeQuery(
+          String jobId,
+          String query,
+          TableReference destinationTable,
+          boolean flattenResults,
+          JobService jobService) throws IOException, InterruptedException {
+        JobConfigurationQuery queryConfig = new JobConfigurationQuery();
+        queryConfig
+            .setQuery(query)
+            .setAllowLargeResults(true)
+            .setCreateDisposition("CREATE_IF_NEEDED")
+            .setDestinationTable(destinationTable)
+            .setFlattenResults(flattenResults)
+            .setPriority("BATCH")
+            .setWriteDisposition("WRITE_EMPTY");
+        jobService.startQueryJob(jobId, queryConfig);
+        Job job = jobService.pollJobStatus(
+            destinationTable.getProjectId(), jobId, MAX_JOB_STATUS_POLL_RETRIES);
+        if (parseStatus(job) != Status.SUCCEEDED) {
+          throw new IOException("Query job failed: " + jobId);
+        }
+        return;
+      }
+
+      private static List<String> executeExtract(
+          String jobId, TableReference table, String tempLocation, JobService jobService)
+              throws InterruptedException, IOException {
+        String uuid = UUID.randomUUID().toString();
+        String destinationUri =
+            String.format("%s/%s/%s", tempLocation, uuid, "*.avro");
+
+        JobConfigurationExtract extract = new JobConfigurationExtract();
+        extract.setSourceTable(table);
+        extract.setDestinationFormat("AVRO");
+        extract.setDestinationUris(ImmutableList.of(destinationUri));
+
+        jobService.startExtractJob(jobId, extract);
+        Job job =
+            jobService.pollJobStatus(table.getProjectId(), jobId, MAX_JOB_STATUS_POLL_RETRIES);
+        if (parseStatus(job) != Status.SUCCEEDED) {
+          throw new IOException("Extract job failed: " + jobId);
+        }
+
+        JobStatistics jobStats = job.getStatistics();
+        LOG.info("JobStats: " + jobStats);
+        List<Long> counts = jobStats.getExtract().getDestinationUriFileCounts();
+        checkState(counts.size() == 1, "Expect one count for the destination uri.");
```

Comment:
these checks would be more helpful if you also print the value you got, and also likely print the first 2/3 if there are more than 1.